packages/core/src/browser/navigatable.ts

```diff
@@ -15,18 +15,59 @@
  ********************************************************************************/
 
 import URI from '../common/uri';
+import { BaseWidget } from './widgets';
+import { WidgetOpenHandler, WidgetOpenerOptions } from './widget-open-handler';
 
 /**
  * Each widget which holds an uri to a workspace file and wants to be able to reveal that file in navigator,
  * (e.g. editor, image viewer, diff editor, etc.) has to implement this interface and provide the file uri on demand.
  * No additional registration is needed.
  */
 export interface Navigatable {
+    /**
+     * Return an underlying file uri.
+     */
     getTargetUri(): URI | undefined;
+    /**
+     * Return a source uri for the given file URI.
```

Comment:
After reading the PR it is still not clear what the purpose of this method is. It seems to be a way to compute a URI based on a file move event. But why can'T we just use the targetURI from the event? Why does it need to get the scheme from the widget?

Reply:
An idea was to have an inverse operation to `getTargetUri`:
- `getTargetUri` gets an underlying file URI from the widget URI, e.g. from a diff URI for the editor widget.
- `getSourceUri` creates a new widget URI based on a given file URI and a current widget URI.

In the case of the diff editor it creates a diff URI. In the case of the preview, it preserves a query if it was opened via a link. A target URI from a move event does not have this information. 

I've added it because it would be correct to replace only a path of URI, but it also makes following (esoteric) cases possible:
![source](https://user-images.githubusercontent.com/3082655/44004279-5e3bf7f8-9e60-11e8-8b34-3d52db458480.gif)

I am not sure about naming. `getTargetUri` was named like that to avoid colliding with `getUri` which is different. `getSourceUri` just to have a symmetric name.

Reply:
Understood, in that case the name should be something like `createMoveToURI` and have the explanation above as its doc comment :)